File: test/document/keys_at_test.rb
# frozen_string_literal: true

require "test_helper"

class DocumentKeysAtTest < Minitest::Spec
  test "keys_at returns the keys of the document root" do
    document = Yerba::Document.parse(<<~YAML)
      name: Conf
      year: 2024
      venue:
        city: Berlin
    YAML

    assert_equal ["name", "year", "venue"], document.keys_at("")
  end

  test "keys_at returns the keys of a nested map" do
    document = Yerba::Document.parse(<<~YAML)
      database:
        host: localhost
        port: 5432
    YAML

    assert_equal ["host", "port"], document.keys_at("database")
  end

  test "keys_at does not descend into a nested sequence of maps" do
    document = Yerba::Document.parse(<<~YAML)
      id: "aloha"
      name: "Aloha"
      channels:
        - id: "UC123"
          handle: "@confreaks"
    YAML

    assert_equal ["id", "name", "channels"], document.keys_at("")
  end

  test "keys_at does not descend into a nested sequence of scalars" do
    document = Yerba::Document.parse(<<~YAML)
      name: Conf
      tags:
        - ruby
        - rails
    YAML

    assert_equal ["name", "tags"], document.keys_at("")
  end

  test "keys_at reports a key whose value is empty" do
    document = Yerba::Document.parse(<<~YAML)
      a: 1
      b:
      c: 3
    YAML

    assert_equal ["a", "b", "c"], document.keys_at("")
  end

  test "keys_at reads the keys of a flow map" do
    document = Yerba::Document.parse("venue: { city: Berlin, country: DE }\n")

    assert_equal ["city", "country"], document.keys_at("venue")
  end

  test "keys_at reads the keys of a map inside a sequence" do
    document = Yerba::Document.parse(<<~YAML)
      - id: "a"
        title: "First"
      - id: "b"
    YAML

    assert_equal ["id", "title"], document.keys_at("[0]")
  end

  test "keys_at returns nothing for a sequence" do
    document = Yerba::Document.parse(<<~YAML)
      tags:
        - ruby
    YAML

    assert_empty document.keys_at("tags")
  end

  test "keys_at returns nothing for a sequence document root" do
    document = Yerba::Document.parse(<<~YAML)
      - id: "a"
      - id: "b"
    YAML

    assert_empty document.keys_at("")
  end

  test "keys_at returns nothing for a scalar" do
    assert_empty Yerba::Document.parse("name: Conf\n").keys_at("name")
  end

  test "keys_at returns nothing for a selector that matches nothing" do
    assert_empty Yerba::Document.parse("name: Conf\n").keys_at("missing")
  end

  test "keys_at matches the keys of the parsed value" do
    document = Yerba::Document.parse(<<~YAML)
      name: Conf
      venue:
        city: Berlin
        country: DE
    YAML

    ["", "venue"].each do |selector|
      assert_equal document.value_at(selector).keys, document.keys_at(selector), selector
    end
  end
end

File: rust/tests/keys_at_test.rs
mod support;
use indoc::indoc;
use support::parse;

#[test]
fn test_keys_at_root() {
  let document = parse(indoc! {"
    name: Conf
    year: 2024
    venue:
      city: Berlin
  "});

  assert_eq!(document.keys_at(""), vec!["name", "year", "venue"]);
}

#[test]
fn test_keys_at_nested_map() {
  let document = parse(indoc! {"
    database:
      host: localhost
      port: 5432
  "});

  assert_eq!(document.keys_at("database"), vec!["host", "port"]);
}

#[test]
fn test_keys_at_does_not_descend_into_nested_sequence() {
  let document = parse(indoc! {"
    id: aloha
    name: Aloha
    channels:
      - id: UC123
        handle: '@confreaks'
  "});

  assert_eq!(document.keys_at(""), vec!["id", "name", "channels"]);
}

#[test]
fn test_keys_at_reports_key_without_a_value() {
  let document = parse(indoc! {"
    a: 1
    b:
    c: 3
  "});

  assert_eq!(document.keys_at(""), vec!["a", "b", "c"]);
}

#[test]
fn test_keys_at_reads_flow_map() {
  let document = parse("venue: { city: Berlin, country: DE }");

  assert_eq!(document.keys_at("venue"), vec!["city", "country"]);
}

#[test]
fn test_keys_at_map_inside_sequence() {
  let document = parse(indoc! {"
    - id: a
      title: First
    - id: b
  "});

  assert_eq!(document.keys_at("[0]"), vec!["id", "title"]);
}

#[test]
fn test_keys_at_returns_nothing_for_a_sequence() {
  let document = parse(indoc! {"
    tags:
      - ruby
  "});

  assert!(document.keys_at("tags").is_empty());
}

#[test]
fn test_keys_at_returns_nothing_for_a_sequence_root() {
  let document = parse(indoc! {"
    - id: a
    - id: b
  "});

  assert!(document.keys_at("").is_empty());
}

#[test]
fn test_keys_at_returns_nothing_for_a_scalar() {
  let document = parse("name: Conf");

  assert!(document.keys_at("name").is_empty());
}

#[test]
fn test_keys_at_returns_nothing_for_a_missing_selector() {
  let document = parse("name: Conf");

  assert!(document.keys_at("missing").is_empty());
}
